doc/GUIDE.md

```diff
@@ -0,0 +1,822 @@
+only available on Windows).
+
+### Functions
+
+Let's say we have a main library header file `mylib.h` that contains the following
+function declaration:
+
+```c
+// mylib.h
+
+void foo(int x);
+```
+
+Jextract will generate the following set of methods for this function, in the main header
+class it generates:
+
+```java
+// mylib_h.java
+
+public static void foo(int x) { ... } // 1
+
+public static MemorySegment foo$address() { ... } // 2
+public static FunctionDescriptor foo$descriptor() { ... } // 3
+public static MethodHandle foo$handle() { ... } // 4
+```
+
+First and foremost, there is a static wrapper method that is generated that can be used to
+call the C function (1). Besides that, there are also several accessors that return
+additional meta-data for the method: the function's address (2), the function descriptor
+(3), and the method handle returned by the FFM linker (4), which is used to implement the
+static wrapper method (1).
+
+The parameter types and return type of this method depend on the carrier types of the
+layouts that make up the function descriptor of the function, which is itself derived
+from the parsed header files.
+
+### Global Variables
+
+For a global variable declaration in a header file like this:
+
+```c
+// mylib.h
+
+int bar;
+```
+
+Jextract generates the following:
+
+```java
+// mylib_h.java
+
+public static OfInt bar$layout() { ... } // 1
+public static MemorySegment bar$segment() { ... } // 2
+
+public static int bar() { ... } // 3
+public static void bar(int varValue) { ... } // 3
+```
+
+`bar$layout` is the FFM memory layout of the global variable (1), and `bar$segment` is the
+address of the global variable (2).
+
+Besides that, jextract also generates a getter and a setter method to get and set the value
+of the global variable (3). Once again, the parameter and return type of the getter and
+setter depend on the carrier type of the layout of the global variable, which is itself
+derived from the header files.
+
+### Constants (Macros & Enums)
+
+Both macros and enums are translated similarly. If we have a header file containing these
+declarations:
+
+```c
+// mylib.h
+
+#define MY_MACRO 42
+enum MY_ENUM { A, B, C };
+```
+
+Jextract will generate a set of simple getter methods to access the constant values of the
+macro and the enum constants:
+
+```java
+// mylib_h.java
+
+public static int MY_MACRO() { ... }
+
+public static int A() { ... }
+public static int B() { ... }
+public static int C() { ... }
+```
+
+Note that the enum constants are exposed as top-level methods, rather than being nested
+inside a class called `MY_ENUM`, or through the use of a Java `enum`. This translation
+strategy mimics C's behavior of enum constants being accessible as a top-level declaration
+as well.
+
+Not all types of macros are supported though. Only macros that have a primitive numerical
+value, a string, or a pointer type are supported. Most notably, function-like macros are
```

Comment:
Several people wonder what to do when dealing with a library that is heavy in its use of function-like macros. There's two strategies here - the first is to replicate the code in the function-like macro in Java, using FFM. The second is to create a small C library which turns the function-like macro into a proper function, so that it can be used from FFM. Not sure we want to add this detail here (or maybe in an appendix).

Reply:
I think something like that is good to add here. This guide is meant to be comprehensive. If we think some information is useful in the use of jextract, let's add it to the guide.